File: c++/string.cpp
// :specimen: string.basics //
	// :basics: Set a variable to a string value //
		#include <string>
		std::string foo = "foo";
		std::string bar("bar");
	// end //
	
	// :basics: Escape a single quote //
		std::string quote = "\'quote\'";
	// end //
	
	// :basics: Multiline strings //
		// Not supported, but you can fake it (badly) with concatenation:
		std::string multi = "a string split " +
		        std::string("across multiple lines ") +
		        std::string("by discretely typing each line";
	// end //
// end //

// :specimen: string.substrings //
	// :substrings: Retrieve parts of a string //
		std::string str = "foobarbaz";
		std::string sub = foo.substr(3,3); // => "bar"
	// end //
	
	// :substrings: Modify parts of a string //
		// Complicated.
		#include <cstring>
		std::string orig = "cstrings are hard";
		strcpy( ((char *)orig.data()) + 13, "easy"); // => "cstrings are easy"
	// end //
// end //

// :specimen: string.defaults //
	// :defaults: Set string defaults //
		
	// end //
	
	// :defaults: Set strings if not already set //
		std::string def;
		if(def.empty())
		  def = "foo";
	// end //
// end //

// :specimen: string.general //
	// :general: Reversing a string //
    std::string abc = "abcdef";
    for(int i=0;i<abc.length()/2;i++) {
      char t = abc.at(i);
      abc[i] = abc.at(abc.length()-i-1);
      abc[abc.length()-i-1] = t;
    }		
	// end //
	
	// :general: String case //
		std::string up = "aAbB";
		for(int i=0;i<up.length();i++) {
		  if(up.at(i) <= 'z' and up.at(i) >= 'a')
		    up[i] += 'A'-'a';
		} // => AABB
		
		std::string down = "aAbB";
		for(int i=0;i<up.length();i++) {
		  if(up.at(i) <= 'Z' and up.at(i) >= 'A')
		    up[i] -= 'A'-'a';
		} // => aabb
		
		std::string word = "word";
		if(word.at(0) <= 'z' and word.at(0) >= 'a')
		  word[0] += 'A'-'a';
		// => Word
	// end //
	
	// :general: Trimming strings //
		
	// end //
// end //

File: ruby/string.rb
# :specimen: string.basics #
	# :basics: Set a variable to a string value #
		string = 'foo bar'
		string = %q/foo bar/
		string = %q[foo bar]
		string = %q{foo bar}
		string = %q(foo bar)
		string = %q<foo bar>
	# end #
	
	# :basics: Escape a single quote #
		string = 'fo\'o b\'ar'
	# end #
	
	# :basics: Multiline strings #
		string = <<"EOF"
		Multi
		line
		terminated by 'EOF' on a line by itself
EOF
	# end #
# end #

# :specimen: string.substrings #
	# :substrings: Retrieve parts of a string #
		string = 'foo bar, this is an example string'
		string[0, 1] # => "f"
		string[5, 2] # => "ar"
		string[15..-1] # => "s an example string"
		string[-1, 1] # => "g"
		string[-6..-1] # => "string"
		string[-5, 3] # => "tri"
	# end #
	
	# :substrings: Modify parts of a string #
		string[0, 3] = 'bar' # => "bar bar, this is an example string"
		string[0, 1] = '' # => "ar bar, this is an example string"
	# end #
# end #

# :specimen: string.defaults #
	# :defaults: Set string defaults #
		boolean = false
		string = boolean || 'foo' # => "foo"
		boolean = true
		string = boolean || 'foo' # => true
	# end #
	
	# :defaults: Set strings if not already set #
		string = nil
		string ||= 'foo' # => "foo"
		string ||= 'bar' # => "foo"
	# end #
# end #

# :specimen: string.general #
	# :general: Reversing a string #
		string = 'foo'
		string.reverse # => "oof"
	# end #
	
	# :general: String case #
		string = 'foo'
		string.upcase # => "FOO"
		string.downcase # => "foo"
		string.capitalize # => "Foo"
	# end #
	
	# :general: Trimming strings #
		string = '    foobar '
		string.strip # => "foobar"
	# end #
# end #
